### PR TITLE
90%: Metatron Manifestation Support

### PR DESCRIPTION
**Pull request by rsinger**
This PR introduces `Talis::Bibliography::Manifestation` (a DSL for Metatron Manifestations API) and the ability to specify the Persona host for specific services (the use case here is that Metatron *must* be hosted, so the ability to allow some services - e.g. Blueprint - to use local Persona to not interfere with other services - e.g. Metatron - using staging Persona, etc.).

TODO:

- [x] Tests

### CODE REVIEW

**Comment by omarqureshi** [comment]
Looks good, only thing I can probably say is that if you ever add any more of these pluggable overrides, we should group them together in the same directory. Otherwise, good job!

**Comment by rsinger** [comment]
@omarqureshi Do you know of a DRYer way to automatically extend these classes?  That is, I wanted to extend from (e.g.) Talis::Bibliography or Talis::Hierarchy, but I couldn't figure out a way to make that matriculate to Talis::Bibliography::Work or Talis::Hierarchy::Node, etc.

**Comment by rsinger** [comment]
@omarqureshi - could you review this for me?

File: lib/talis/bibliography.rb [hunk 1]
```diff
@@ -1,10 +1,44 @@
...
+    end
+
+    private
+
+    def configure_metatron
+      MetatronClient.configure do |config|
+        config.scheme = base_uri[/https?/]
+        config.host = base_uri
+        # Non-production environments have a base path
+        if ENV['METATRON_BASE_PATH']
```
Comment (omarqureshi) [inline lib/talis/bibliography.rb]: does this matter if we set base_path to null?

Comment (rsinger) [inline lib/talis/bibliography.rb]: metatron_rb has a default base_path of `/2/`

Comment (omarqureshi) [inline lib/talis/bibliography.rb]: ok - great, leave as is

File: lib/talis/bibliography/manifestation.rb [hunk 2]
```diff
@@ -0,0 +1,142 @@
...
+          if rel == :contributors
+            add_related_contributors(manifestation_data)
+          else
+            add_related_work(manifestation_data)
+          end
+        end
+      end
+
+      def add_related_contributors(manifestation_data)
+        manifestation_data.relationships.contributors.data.each do |resource|
```
Comment (omarqureshi) [inline lib/talis/bibliography/manifestation.rb]: can you not just append to contributors?

```ruby
contributors += manifestation_data.relationships.contributors.data
```

File: lib/talis/bibliography/manifestation.rb [hunk 3]
```diff
@@ -0,0 +1,142 @@
...
+        included.find do |resource|
+          resource.id == resource_data.id && resource.type == resource_data.type
+        end
+      end
+
+      def parse_manifestation_data(manifestation_data)
+        @manifestation_data = manifestation_data
+        @title = manifestation_data.try(:attributes).try(:title)
+
+        unless manifestation_data.relationships.nil?
```
Comment (omarqureshi) [inline lib/talis/bibliography/manifestation.rb]: if any?

File: lib/talis/bibliography/manifestation.rb [hunk 4]
```diff
@@ -0,0 +1,142 @@
...
+        end
+      end
+
+      def add_related_contributors(manifestation_data)
+        manifestation_data.relationships.contributors.data.each do |resource|
+          contributors << resource
+        end
+      end
+
+      def add_related_work(manifestation_data)
```
Comment (omarqureshi) [inline lib/talis/bibliography/manifestation.rb]: is there only ever one related work?

Comment (omarqureshi) [inline lib/talis/bibliography/manifestation.rb]: ok - i see where it's called from

File: lib/talis/bibliography/work.rb [hunk 5]
```diff
@@ -127,13 +97,14 @@ def assets
       # typed objects
       # @param resources [Array] an array of Metatron::ResourceData objects
       def hydrate_relationships(resources)
-        manifestations.each_with_index do |manifestation, idx|
-          resource = find_relationship_in_included manifestation.to_hash,
+        manifestations.each_with_index do |m, idx|
```
Comment (omarqureshi) [inline lib/talis/bibliography/work.rb]: can this not be written as a #map ?

```ruby
manifestations.map do |m|
  resource = find_relationship_in_included m.manifestation_data.to_hash, resources
  if resource
    Manifestation.new(MetatronClient::ManifestationData.new(resource.to_hash))
  else
    m
  end
end
```

File: lib/talis/bibliography/manifestation.rb (hunk 4 above)
Comment (omarqureshi) [inline lib/talis/bibliography/manifestation.rb]: I suppose you're protected by the fact that this is a private method call - but, should probably be careful here if for some reason there would ever be more than one work.

File: lib/talis/bibliography/manifestation.rb [hunk 6]
```diff
@@ -0,0 +1,142 @@
...
+        @assets ||= []
+      end
+
+      # By default, the metatron client returns generic ResourceLink objects
+      # as the related resources.  When passed an array of Metatron::ResourceData
+      # objects, it will replace the ResourceLink objects with more appropriately
+      # typed objects
+      # @param resources [Array] an array of Metatron::ResourceData objects
+      def hydrate_relationships(included_resources)
+        contributors.each_with_index do |contributor, idx|
```
Comment (omarqureshi) [inline lib/talis/bibliography/manifestation.rb]: can #map here too

File: lib/talis/bibliography/manifestation.rb [hunk 7]
```diff
@@ -0,0 +1,142 @@
...
+        #   server.
+        # @raise [Talis::Errors::ServerCommunicationError] for network issues.
+        def find(request_id: new_req_id, opts: {})
+          api_client(request_id).manifestation(token, opts)
+                                .extend(ResultSet).hydrate
+        rescue MetatronClient::ApiError => error
+          begin
+            handle_response(error)
+          rescue Talis::Errors::NotFoundError
+            empty_result_set(MetatronClient::ManifestationResultSet, count: 0)
```
Comment (omarqureshi) [inline lib/talis/bibliography/manifestation.rb]: I like this

File: lib/talis/bibliography/manifestation.rb (hunk 4 above)
Comment (rsinger) [inline lib/talis/bibliography/manifestation.rb]: There can't be more than one work.  That's defined in the swagger doc.

File: lib/talis/bibliography/work.rb (hunk 5 above)
Comment (rsinger) [inline lib/talis/bibliography/work.rb]: Actually, this method's return is fairly irrelevant, it really should be returning void.  Do you still think it should be a map in that scenario?

Comment (omarqureshi) [inline lib/talis/bibliography/work.rb]: No - should be an each (without index)

File: lib/talis/bibliography/manifestation.rb (hunk 3 above)
Comment (rsinger) [inline lib/talis/bibliography/manifestation.rb]: Unfortunately, this is either `nil` or a `MetatronClient::ManifestationRelationships` object (which doesn't implement `.any()`), so I think this is as good as we're going to get.

**Pull request merged**